src/core/documentation/documentation.ts

```diff
@@ -0,0 +1,158 @@
+        format: any,
+        content: string
+    ) {
+        if (element && !this.model.containsElement(element)) {
+            throw new Error(
+                `The element named ${element.name} does not exist in the model associated with this documentation.`
+            );
+        }
+
+        assert(title, "A title must be specified.");
+        assert(content, "Content must be specified.");
+        assert(format, "A format must be specified.");
+        this.checkSectionIsUnqiue(title, element);
+
+        const section = new Section();
+        section.fromDto({
+            element,
+            title,
+            order: this.sections.size + 1,
+            format,
+            content
+        });
+        this.sections.add(section);
+        return section;
+    }
+
+    public addDecision(
+        element: SoftwareSystem,
+        id: string,
+        date: Date,
+        title: string,
+        status: DecisionStatus,
+        format: Format,
+        content: string
+    ) {
+        assert(id, `An id must be specifed`);
+        assert(date, "A date must be specified");
+        assert(title, "A title must be specified");
+        assert(status, "A status must be specified");
+        assert(format, "A format must be specified");
+        this.checkDecisionIsUnqiue(id, element);
+
+        const decision = new Decision();
+        decision.fromDto({ id, date, title, status, format, content });
+        this.decisions.add(decision);
+        return decision;
+    }
+
+    public toDto() {
+        return {
+            sections: Array.from(this.sections.values()).map(section =>
+                section.toDto()
+            ),
+            decisions: Array.from(this.decisions.values()).map(decision =>
+                decision.toDto()
+            )
+        };
+    }
+
+    public fromDto(dto: any) {
+        if (dto.sections) {
+            dto.sections.forEach((sectionDto: any) => {
+                const section = new Section();
+                section.fromDto(sectionDto);
+                this.sections.add(section);
+            });
+        }
+        if (dto.decisions) {
+            dto.decisions.forEach((decisionDto: any) => {
+                const decision = new Decision();
+                decision.fromDto(decisionDto);
+                this.decisions.add(decision);
+            });
+        }
+    }
+
+    public hydrate() {
+        this.sections.forEach(section => this.findElementAndHydrate(section));
+        this.decisions.forEach(decision =>
+            this.findElementAndHydrate(decision)
+        );
+    }
+
+    private findElementAndHydrate(documentationElement: {
+        elementId?: string;
+        element?: Element;
+    }) {
+        if (documentationElement.elementId) {
+            documentationElement.element = this.model.getElement(
+                documentationElement.elementId
+            );
+        }
+    }
+
+    private checkSectionIsUnqiue(title: string, element?: Element) {
```

Comment:
The previous way I implemented these checks was trying to be too clever and buggy, Instead, I fell back to a straight for of the java client's implementation.